File: apps/ios/NanoLink/Screens/AssistantScreen.swift
import SwiftUI

private struct AssistantTurn: Identifiable {
    let id = UUID()
    let role: String
    let content: String
    let isError: Bool

    var isUser: Bool { role == "user" }
}

/// Combined AI assistant screen with findings + chat.
struct AssistantScreen: View {
    @EnvironmentObject private var store: AppStore
    @Environment(\.nano) private var t

    @State private var findings: [AssistantFinding] = []
    @State private var turns: [AssistantTurn] = []
    @State private var input = ""
    @State private var sending = false
    @FocusState private var focused: Bool

    var body: some View {
        VStack(spacing: 0) {
            ScrollViewReader { proxy in
                ScrollView {
                    LazyVStack(spacing: 12, pinnedViews: []) {
                        if !findings.isEmpty {
                            findingsSection
                        }

                        if turns.isEmpty && !sending {
                            emptyChatState
                        }

                        ForEach(turns) { turn in
                            chatBubble(turn).id(turn.id)
                        }

                        if sending {
                            thinkingBubble.id("thinking")
                        }
                    }
                    .padding(EdgeInsets(top: 12, leading: 16, bottom: 32, trailing: 16))
                }
                .onChange(of: turns.count) { _ in scroll(proxy) }
                .onChange(of: sending) { _ in scroll(proxy) }
            }

            composer
        }
        .background(t.bg.ignoresSafeArea())
        .navigationTitle(tr("assistant.title"))
        .navigationBarTitleDisplayMode(.inline)
        .task { await loadFindings() }
    }

    private var findingsSection: some View {
        VStack(alignment: .leading, spacing: 8) {
            Text(tr("assistant.diagnosis"))
                .font(.system(size: 12, weight: .semibold))
                .foregroundColor(t.fg3)

            FlowLayout(spacing: 8) {
                ForEach(findings) { finding in
                    findingCard(finding)
                }
            }
        }
    }

    private func findingCard(_ finding: AssistantFinding) -> some View {
        let (icon, tone) = toneIcon(finding.kind)

        return NanoCard {
            VStack(alignment: .leading, spacing: 8) {
                HStack(spacing: 8) {
                    Image(systemName: icon)
                        .font(.system(size: 16))
                        .foregroundColor(tone)
                    Text(finding.title)
                        .font(.system(size: 14, weight: .semibold))
                        .foregroundColor(t.fg)
                }

                Text(finding.detail)
                    .font(.system(size: 12))
                    .foregroundColor(t.fg3)
                    .lineSpacing(4)

                if !finding.actions.isEmpty, let agentId = finding.agentId {
                    FlowLayout(spacing: 6) {
                        ForEach(finding.actions, id: \.self) { action in
                            actionChip(action, agentId: agentId)
                        }
                    }
                }
            }
            .padding(12)
        }
    }

    private func actionChip(_ action: String, agentId: String) -> some View {
        let agent = store.agentById(agentId)
        let initialTab = action == "history" ? 1 : action == "shell" || action == "terminal" ? 2 : 0

        return NavigationLink {
            if let agent = agent {
                AgentDetailScreen(agent: agent, initialTab: initialTab)
            }
        } label: {
            Text(tr("assistant.action.\(action)"))
                .font(.system(size: 11, weight: .medium))
                .foregroundColor(t.accent)
                .padding(.horizontal, 10)
                .padding(.vertical, 4)
                .background(t.accent.opacity(0.14))
                .clipShape(Capsule())
        }
    }

    private var emptyChatState: some View {
        VStack(spacing: 16) {
            Image(systemName: "sparkles")
                .font(.system(size: 42))
                .foregroundColor(t.fg4)

            Text(tr("assistant.greeting"))
                .font(.system(size: 18, weight: .semibold))
                .foregroundColor(t.fg)

            Text(tr("assistant.intro"))
                .font(.system(size: 13))
                .foregroundColor(t.fg3)
                .multilineTextAlignment(.center)
                .padding(.horizontal, 24)

            Spacer().frame(height: 8)

            Text(tr("assistant.tryThese"))
                .font(.system(size: 12, weight: .semibold))
                .foregroundColor(t.fg3)

            FlowLayout(spacing: 8) {
                ForEach(suggestions, id: \.self) { suggestion in
                    Button { send(suggestion) } label: {
                        Text(suggestion)
                            .font(.system(size: 12))
                            .foregroundColor(t.fg2)
                            .padding(.horizontal, 14)
                            .padding(.vertical, 8)
                            .background(t.card2)
                            .clipShape(Capsule())
                    }
                }
            }
            .padding(.horizontal, 16)
        }
        .padding(.vertical, 32)
    }

    private var suggestions: [String] {
        [
            tr("assistant.suggestion.highCpu"),
            tr("assistant.suggestion.diskSpace"),
            tr("assistant.suggestion.offlineNodes"),
            tr("assistant.suggestion.resourceTrend"),
        ]
    }

    private func chatBubble(_ turn: AssistantTurn) -> some View {
        HStack {
            if turn.isUser { Spacer(minLength: 50) }

            Text(turn.content)
                .font(.system(size: 14))
                .foregroundColor(turn.isUser ? t.fg : turn.isError ? t.crit : t.fg2)
                .lineSpacing(5)
                .padding(.horizontal, 14)
                .padding(.vertical, 10)
                .background(turn.isUser ? t.accent.opacity(0.14) : turn.isError ? t.crit.opacity(0.12) : t.card2)
                .clipShape(RoundedRectangle(cornerRadius: 16, style: .continuous))

            if !turn.isUser { Spacer(minLength: 50) }
        }
    }

    private var thinkingBubble: some View {
        HStack {
            HStack(spacing: 6) {
                ForEach(0..<3, id: \.self) { _ in
                    Circle()
                        .fill(t.fg3)
                        .frame(width: 6, height: 6)
                }
            }
            .padding(.horizontal, 14)
            .padding(.vertical, 10)
            .background(t.card2)
            .clipShape(RoundedRectangle(cornerRadius: 16, style: .continuous))

            Spacer(minLength: 50)
        }
    }

    private var composer: some View {
        HStack(alignment: .bottom, spacing: 8) {
            TextField(tr("assistant.inputHint"), text: $input, axis: .vertical)
                .font(.system(size: 14))
                .focused($focused)
                .lineLimit(1...5)
                .padding(.horizontal, 14)
                .padding(.vertical, 10)
                .background(t.card)
                .clipShape(RoundedRectangle(cornerRadius: 20, style: .continuous))
                .submitLabel(.send)
                .onSubmit { send(nil) }
                .disabled(sending)

            Button { send(nil) } label: {
                Image(systemName: "sparkles")
                    .font(.system(size: 20))
                    .foregroundColor(t.onAccent)
            }
            .frame(width: 40, height: 40)
            .background(
                (sending || input.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty) ?
                    t.card2 : t.accent
            )
            .clipShape(Circle())
            .disabled(sending || input.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty)
        }
        .padding(EdgeInsets(top: 10, leading: 16, bottom: 10, trailing: 16))
        .background(t.bg)
    }

    private func scroll(_ proxy: ScrollViewProxy) {
        guard !turns.isEmpty || sending else { return }
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
            withAnimation {
                if sending {
                    proxy.scrollTo("thinking", anchor: .bottom)
                } else {
                    proxy.scrollTo(turns.last?.id, anchor: .bottom)
                }
            }
        }
    }

    private func send(_ preset: String?) {
        let message = preset ?? input.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !message.isEmpty, !sending else { return }

        turns.append(AssistantTurn(role: "user", content: message, isError: false))
        input = ""
        sending = true

        Task {
            guard let serverId = store.activeServerId,
                  let service = store.serviceForServer(serverId) else {
                sending = false
                turns.append(AssistantTurn(role: "assistant", content: tr("assistant.error.notConfigured"), isError: true))
                return
            }

            let history = turns.map { ChatMessage(role: $0.role, content: $0.content) }
            let result = await service.assistantChat(history)

            await MainActor.run {
                sending = false
                if result.ok, let reply = result.reply {
                    turns.append(AssistantTurn(role: reply.role, content: reply.content, isError: false))
                } else {
                    let errorMsg = errorText(result.error)
                    turns.append(AssistantTurn(role: "assistant", content: errorMsg, isError: true))
                }
            }
        }
    }

    private func errorText(_ error: AssistantChatError?) -> String {
        switch error {
        case .notConfigured: return tr("assistant.error.notConfigured")
        case .badRequest: return tr("assistant.error.badRequest")
        case .upstreamFailed: return tr("assistant.error.upstreamFailed")
        case .serverError: return tr("assistant.error.serverError")
        case .network: return tr("assistant.error.network")
        case .none: return tr("assistant.error.unknown")
        }
    }

    private func loadFindings() async {
        guard let serverId = store.activeServerId,
              let service = store.serviceForServer(serverId) else { return }
        if let result = await service.fetchAssistantFindings() {
            await MainActor.run { findings = result }
        }
    }

    private func toneIcon(_ kind: String) -> (String, Color) {
        switch kind {
        case "anomaly": return ("bolt.fill", t.crit)
        case "warn": return ("exclamationmark.triangle.fill", t.warn)
        case "info": return ("info.circle.fill", t.info)
        case "ok": return ("checkmark.circle.fill", t.ok)
        default: return ("info.circle.fill", t.fg3)
        }
    }
}

/// Wrapping layout used by finding actions and suggested prompts.
private struct FlowLayout: Layout {
    var spacing: CGFloat = 8

    func sizeThatFits(
        proposal: ProposedViewSize,
        subviews: Subviews,
        cache: inout ()
    ) -> CGSize {
        let width = proposal.width ?? 0
        var x: CGFloat = 0
        var y: CGFloat = 0
        var rowHeight: CGFloat = 0

        for view in subviews {
            let size = view.sizeThatFits(.unspecified)
            if x + size.width > width, x > 0 {
                x = 0
                y += rowHeight + spacing
                rowHeight = 0
            }
            x += size.width + spacing
            rowHeight = max(rowHeight, size.height)
        }
        return CGSize(width: width, height: y + rowHeight)
    }

    func placeSubviews(
        in bounds: CGRect,
        proposal: ProposedViewSize,
        subviews: Subviews,
        cache: inout ()
    ) {
        var x = bounds.minX
        var y = bounds.minY
        var rowHeight: CGFloat = 0

        for view in subviews {
            let size = view.sizeThatFits(.unspecified)
            if x + size.width > bounds.maxX, x > bounds.minX {
                x = bounds.minX
                y += rowHeight + spacing
                rowHeight = 0
            }
            view.place(at: CGPoint(x: x, y: y), proposal: ProposedViewSize(size))
            x += size.width + spacing
            rowHeight = max(rowHeight, size.height)
        }
    }
}

File: apps/android/app/src/main/java/com/nanolink/app/ui/screens/AssistantScreen.kt
package com.nanolink.app.ui.screens

import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.outlined.ArrowBack
import androidx.compose.material.icons.outlined.AutoAwesome
import androidx.compose.material.icons.outlined.Bolt
import androidx.compose.material.icons.outlined.CheckCircle
import androidx.compose.material.icons.outlined.Error
import androidx.compose.material.icons.outlined.Info
import androidx.compose.material.icons.outlined.Warning
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.Text
import androidx.compose.material3.TextField
import androidx.compose.material3.TextFieldDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.scale
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.nanolink.app.data.model.AssistantChatError
import com.nanolink.app.data.model.AssistantFinding
import com.nanolink.app.data.model.ChatMessage
import com.nanolink.app.state.AppViewModel
import com.nanolink.app.ui.design.NanoCard
import com.nanolink.app.ui.design.NanoMonoFamily
import com.nanolink.app.ui.design.nano
import com.nanolink.app.ui.tr
import kotlinx.coroutines.launch
import java.util.UUID

private data class AssistantTurn(
    val id: String = UUID.randomUUID().toString(),
    val role: String,
    val content: String,
    val isError: Boolean = false,
)

@Composable
fun AssistantScreen(
    viewModel: AppViewModel,
    onBack: () -> Unit,
    onOpenAgent: (String) -> Unit,
    onOpenTerminal: (String) -> Unit,
    modifier: Modifier = Modifier,
) {
    val t = nano
    val scope = rememberCoroutineScope()
    val listState = rememberLazyListState()
    val turns = remember { mutableStateListOf<AssistantTurn>() }
    var findings by remember { mutableStateOf<List<AssistantFinding>>(emptyList()) }
    var inputText by remember { mutableStateOf("") }
    var sending by remember { mutableStateOf(false) }

    fun send(preset: String? = null) {
        val message = preset ?: inputText.trim()
        if (message.isEmpty() || sending) return

        turns.add(AssistantTurn(role = "user", content = message))
        inputText = ""
        sending = true

        scope.launch {
            val history = turns.map { ChatMessage(role = it.role, content = it.content) }
            val service = viewModel.activeServer?.id?.let(viewModel::serviceForServer)
            if (service == null) {
                sending = false
                turns.add(
                    AssistantTurn(
                        role = "assistant",
                        content = viewModel.localization.text("errors.serverNotConnected"),
                        isError = true,
                    ),
                )
                return@launch
            }
            val result = service.assistantChat(history)
            sending = false

            if (result.ok && result.reply != null) {
                turns.add(AssistantTurn(role = result.reply.role, content = result.reply.content))
            } else {
                val errorMsg = when (result.error) {
                    AssistantChatError.NOT_CONFIGURED -> viewModel.localization.text("assistant.error.notConfigured")
                    AssistantChatError.BAD_REQUEST -> viewModel.localization.text("assistant.error.badRequest")
                    AssistantChatError.UPSTREAM_FAILED -> viewModel.localization.text("assistant.error.upstreamFailed")
                    AssistantChatError.SERVER_ERROR -> viewModel.localization.text("assistant.error.serverError")
                    AssistantChatError.NETWORK -> viewModel.localization.text("assistant.error.network")
                    else -> result.message ?: viewModel.localization.text("assistant.error.unknown")
                }
                turns.add(AssistantTurn(role = "assistant", content = errorMsg, isError = true))
            }
        }
    }

    LaunchedEffect(Unit) {
        val result = viewModel.activeServer?.id
            ?.let(viewModel::serviceForServer)
            ?.fetchAssistantFindings()
        if (result != null) findings = result
    }

    LaunchedEffect(turns.size, sending) {
        if (turns.isNotEmpty() || sending) {
            val target = if (sending) turns.size else turns.size - 1
            listState.animateScrollToItem(target.coerceAtLeast(0))
        }
    }

    Column(modifier = modifier.fillMaxSize()) {
        Row(
            modifier = Modifier.fillMaxWidth().padding(start = 4.dp, end = 16.dp, top = 4.dp, bottom = 4.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            IconButton(onClick = onBack) {
                Icon(Icons.AutoMirrored.Outlined.ArrowBack, contentDescription = tr("common.cancel"), tint = t.fg)
            }
            Icon(Icons.Outlined.AutoAwesome, contentDescription = null, tint = t.accent, modifier = Modifier.size(20.dp))
            Spacer(Modifier.width(8.dp))
            Text(
                tr("assistant.title"),
                fontSize = 22.sp,
                fontWeight = FontWeight.SemiBold,
                color = t.fg,
            )
        }

        LazyColumn(
            state = listState,
            modifier = Modifier.weight(1f).fillMaxWidth(),
            contentPadding = PaddingValues(horizontal = 16.dp, vertical = 12.dp),
            verticalArrangement = Arrangement.spacedBy(10.dp),
        ) {
            if (findings.isNotEmpty()) {
                item { FindingsSection(findings, onOpenAgent, onOpenTerminal) }
            }

            if (turns.isEmpty() && !sending) {
                item { EmptyChatState(onSendPreset = ::send) }
            }

            items(turns, key = { it.id }) { turn ->
                ChatBubble(turn)
            }

            if (sending) {
                item { ThinkingBubble() }
            }
        }

        Composer(
            value = inputText,
            onValueChange = { inputText = it },
            onSend = { send() },
            enabled = !sending,
        )
    }
}

@OptIn(ExperimentalLayoutApi::class)
@Composable
private fun FindingsSection(
    findings: List<AssistantFinding>,
    onOpenAgent: (String) -> Unit,
    onOpenTerminal: (String) -> Unit,
) {
    val t = nano
    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
        Text(
            tr("assistant.diagnosis"),
            fontSize = 12.sp,
            fontWeight = FontWeight.SemiBold,
            color = t.fg3,
        )
        FlowRow(
            horizontalArrangement = Arrangement.spacedBy(8.dp),
            verticalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            findings.forEach { finding ->
                FindingCard(finding, onOpenAgent, onOpenTerminal)
            }
        }
    }
}

@OptIn(ExperimentalLayoutApi::class)
@Composable
private fun FindingCard(
    finding: AssistantFinding,
    onOpenAgent: (String) -> Unit,
    onOpenTerminal: (String) -> Unit,
) {
    val t = nano
    val (icon, tone) = when (finding.kind) {
        "anomaly" -> Icons.Outlined.Bolt to t.crit
        "warn" -> Icons.Outlined.Warning to t.warn
        "info" -> Icons.Outlined.Info to t.info
        "ok" -> Icons.Outlined.CheckCircle to t.ok
        else -> Icons.Outlined.Info to t.fg3
    }

    NanoCard(
        modifier = Modifier.fillMaxWidth(),
        padding = PaddingValues(12.dp),
    ) {
        Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Icon(icon, contentDescription = null, tint = tone, modifier = Modifier.size(16.dp))
                Spacer(Modifier.width(8.dp))
                Text(finding.title, fontSize = 14.sp, fontWeight = FontWeight.SemiBold, color = t.fg)
            }
            Text(finding.detail, fontSize = 12.sp, color = t.fg3, lineHeight = 16.sp)

            if (finding.actions.isNotEmpty() && finding.agentId != null) {
                FlowRow(
                    horizontalArrangement = Arrangement.spacedBy(6.dp),
                    verticalArrangement = Arrangement.spacedBy(6.dp),
                ) {
                    finding.actions.forEach { action ->
                        ActionChip(
                            label = tr("assistant.action.$action"),
                            onClick = {
                                when (action) {
                                    "history" -> onOpenAgent(finding.agentId)
                                    "shell", "terminal" -> onOpenTerminal(finding.agentId)
                                    else -> onOpenAgent(finding.agentId)
                                }
                            },
                        )
                    }
                }
            }
        }
    }
}

@Composable
private fun ActionChip(label: String, onClick: () -> Unit) {
    val t = nano
    Box(
        modifier = Modifier
            .height(24.dp)
            .clip(RoundedCornerShape(999.dp))
            .background(t.accent.copy(alpha = 0.14f))
            .clickable(onClick = onClick)
            .padding(horizontal = 10.dp),
        contentAlignment = Alignment.Center,
    ) {
        Text(
            label,
            fontSize = 11.sp,
            fontWeight = FontWeight.Medium,
            color = t.accent,
        )
    }
}

@OptIn(ExperimentalLayoutApi::class)
@Composable
private fun EmptyChatState(onSendPreset: (String) -> Unit) {
    val t = nano
    val suggestions = listOf(
        tr("assistant.suggestion.highCpu"),
        tr("assistant.suggestion.diskSpace"),
        tr("assistant.suggestion.offlineNodes"),
        tr("assistant.suggestion.resourceTrend"),
    )

    Column(
        modifier = Modifier.fillMaxWidth().padding(vertical = 32.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(16.dp),
    ) {
        Icon(Icons.Outlined.AutoAwesome, contentDescription = null, tint = t.fg4, modifier = Modifier.size(42.dp))
        Text(tr("assistant.greeting"), fontSize = 18.sp, fontWeight = FontWeight.SemiBold, color = t.fg)
        Text(
            tr("assistant.intro"),
            fontSize = 13.sp,
            color = t.fg3,
            modifier = Modifier.padding(horizontal = 24.dp),
        )

        Spacer(Modifier.height(8.dp))
        Text(tr("assistant.tryThese"), fontSize = 12.sp, fontWeight = FontWeight.SemiBold, color = t.fg3)

        FlowRow(
            modifier = Modifier.padding(horizontal = 16.dp),
            horizontalArrangement = Arrangement.spacedBy(8.dp),
            verticalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            suggestions.forEach { suggestion ->
                Box(
                    modifier = Modifier
                        .clip(RoundedCornerShape(999.dp))
                        .background(t.card2)
                        .clickable { onSendPreset(suggestion) }
                        .padding(horizontal = 14.dp, vertical = 8.dp),
                ) {
                    Text(suggestion, fontSize = 12.sp, color = t.fg2)
                }
            }
        }
    }
}

@Composable
private fun ChatBubble(turn: AssistantTurn) {
    val t = nano
    val isUser = turn.role == "user"
    val bgColor = if (isUser) t.accent.copy(alpha = 0.14f) else if (turn.isError) t.crit.copy(alpha = 0.12f) else t.card2
    val textColor = if (isUser) t.fg else if (turn.isError) t.crit else t.fg2

    Row(
        modifier = Modifier.fillMaxWidth(),
        horizontalArrangement = if (isUser) Arrangement.End else Arrangement.Start,
    ) {
        Box(
            modifier = Modifier
                .clip(RoundedCornerShape(16.dp))
                .background(bgColor)
                .padding(horizontal = 14.dp, vertical = 10.dp),
        ) {
            Text(
                turn.content,
                fontSize = 14.sp,
                color = textColor,
                lineHeight = 19.sp,
            )
        }
    }
}

@Composable
private fun ThinkingBubble() {
    val t = nano
    val transition = rememberInfiniteTransition(label = "thinking")
    val progress = transition.animateFloat(
        initialValue = 0f,
        targetValue = 1f,
        animationSpec = infiniteRepeatable(tween(1_200, easing = LinearEasing), RepeatMode.Restart),
        label = "thinkingProgress",
    ).value

    Row(
        modifier = Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.Start,
    ) {
        Box(
            modifier = Modifier
                .clip(RoundedCornerShape(16.dp))
                .background(t.card2)
                .padding(horizontal = 14.dp, vertical = 10.dp),
        ) {
            Row(horizontalArrangement = Arrangement.spacedBy(6.dp)) {
                repeat(3) { index ->
                    Box(
                        modifier = Modifier
                            .size(6.dp)
                            .scale(1f + 0.4f * kotlin.math.sin((progress + index * 0.33f) * 2 * kotlin.math.PI.toFloat()))
                            .alpha(0.4f + 0.4f * kotlin.math.sin((progress + index * 0.33f) * 2 * kotlin.math.PI.toFloat()))
                            .background(t.fg3, CircleShape),
                    )
                }
            }
        }
    }
}

@Composable
private fun Composer(
    value: String,
    onValueChange: (String) -> Unit,
    onSend: () -> Unit,
    enabled: Boolean,
) {
    val t = nano
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .background(t.bg)
            .padding(horizontal = 16.dp, vertical = 10.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        TextField(
            value = value,
            onValueChange = onValueChange,
            placeholder = { Text(tr("assistant.inputHint"), color = t.fg4, fontSize = 14.sp) },
            colors = TextFieldDefaults.colors(
                focusedContainerColor = t.card,
                unfocusedContainerColor = t.card,
                disabledContainerColor = t.card,
                focusedIndicatorColor = Color.Transparent,
                unfocusedIndicatorColor = Color.Transparent,
                disabledIndicatorColor = Color.Transparent,
            ),
            shape = RoundedCornerShape(20.dp),
            modifier = Modifier.weight(1f),
            enabled = enabled,
        )
        Box(
            modifier = Modifier
                .size(40.dp)
                .clip(CircleShape)
                .background(if (enabled && value.trim().isNotEmpty()) t.accent else t.card2)
                .clickable(enabled = enabled && value.trim().isNotEmpty(), onClick = onSend),
            contentAlignment = Alignment.Center,
        ) {
            if (enabled && value.trim().isNotEmpty()) {
                Icon(
                    Icons.Outlined.AutoAwesome,
                    contentDescription = tr("common.send"),
                    tint = t.onAccent,
                    modifier = Modifier.size(20.dp),
                )
            } else {
                CircularProgressIndicator(
                    color = t.fg4,
                    strokeWidth = 2.dp,
                    modifier = Modifier.size(18.dp).alpha(if (enabled) 0f else 1f),
                )
            }
        }
    }
}
